src/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs

```diff
@@ -211,5 +205,99 @@ private Socket GetOrCreateAcceptSocket(Socket acceptSocket, bool checkDisconnect
             handle = acceptSocket._handle;
             return acceptSocket;
         }
+
+        private void SendFileInternal(string fileName, byte[] preBuffer, byte[] postBuffer, TransmitFileOptions flags)
+        {
+            // Open the file, if any
+            FileStream fileStream = null;
+            if (fileName != null && fileName.Length > 0)
+            {
+                fileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read);
+            }
+
+            SocketError errorCode;
+            using (fileStream)
+            {
+                SafeFileHandle fileHandle = fileStream?.SafeFileHandle;
+
+                // This can throw ObjectDisposedException.
+                errorCode = SocketPal.SendFile(_handle, fileHandle, preBuffer, postBuffer, flags);
+            }
+
+            if (errorCode != SocketError.Success)
+            {
+                SocketException socketException = new SocketException((int)errorCode);
+                UpdateStatusAfterSocketError(socketException);
+                if (NetEventSource.IsEnabled) NetEventSource.Error(this, socketException);
+                throw socketException;
+            }
+
+            // If the TransmitFile call did a Disconnect, process that now
+            if ((flags & (TransmitFileOptions.Disconnect | TransmitFileOptions.ReuseSocket)) != 0)
```

Comment:
The comment is confusing.  The user can pass in these flags to cause TransmitFile to do a disconnect/reuse when it's done.  If they did pass these flags, then we want to update the state of the managed Socket to reflect this.

TransmitFileOptions.ReuseSocket is strictly a Windows thing.  We could probably support TransmitFileOptions.Disconnect here, but I'm not sure it's worth it.  I based the behavior here on how we handle Socket.Disconnect currently -- that is, it's not supported on Unix, even in the non-reuse case.  See https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L1326.